core/lib/l1_contract_interface/src/i_executor/structures/commit_batch_info.rs

```diff
@@ -165,56 +162,219 @@ impl<'a> CommitBatchInfo<'a> {
+                    // We compute and add the blob commitment to the pubdata payload so that we can verify the proof
+                    // even if we are not using blobs.
+                    let blob_commitment = KzgInfo::new(&pubdata).to_blob_commitment();
 
-        let reference = match reference {
-            Token::Tuple(tuple) => tuple,
-            _ => {
-                return Err(parse_error(format!(
-                    "reference has unexpected shape; expected a tuple, got {reference:?}"
-                )))
+                    let result = std::iter::once(PUBDATA_SOURCE_CALLDATA)
+                        .chain(pubdata)
+                        .chain(blob_commitment)
+                        .collect();
+
+                    tokens.push(Token::Bytes(result));
+                }
+                PubdataDA::Blobs => {
+                    let pubdata_commitments =
+                        pubdata.chunks(ZK_SYNC_BYTES_PER_BLOB).flat_map(|blob| {
+                            let kzg_info = KzgInfo::new(blob);
+                            kzg_info.to_pubdata_commitment()
+                        });
+                    let result = std::iter::once(PUBDATA_SOURCE_BLOBS)
+                        .chain(pubdata_commitments)
+                        .collect();
+
+                    tokens.push(Token::Bytes(result));
+                }
             }
-        };
-        let Some(last_reference_token) = reference.last() else {
-            return Err(parse_error("reference commitment data is empty"));
-        };
+        }
 
-        let last_reference_token = match last_reference_token {
-            Token::Bytes(bytes) => bytes,
-            _ => return Err(parse_error(format!(
-                "last reference token has unexpected shape; expected bytes, got {last_reference_token:?}"
-            ))),
-        };
-        match last_reference_token.first() {
-            Some(&byte) if byte == PUBDATA_SOURCE_CALLDATA => Ok(PubdataDA::Calldata),
-            Some(&byte) if byte == PUBDATA_SOURCE_BLOBS => Ok(PubdataDA::Blobs),
-            Some(&byte) => Err(parse_error(format!(
-                "unexpected first byte of the last reference token; expected one of [{PUBDATA_SOURCE_CALLDATA}, {PUBDATA_SOURCE_BLOBS}], \
-                 got {byte}"
-            ))),
-            None => Err(parse_error("last reference token is empty")),
+        Token::Tuple(tokens)
+    }
+}
+
+/// Encoding for `CommitBatchInfo` from `IExecutor.sol` for a contract running in validium mode.
+#[derive(Debug)]
+pub struct CommitBatchInfoValidium<'a> {
+    pub l1_batch_with_metadata: &'a L1BatchWithMetadata,
+    pub pubdata_da: PubdataDA,
+}
+
+impl<'a> CommitBatchInfoValidium<'a> {
+    pub fn new(l1_batch_with_metadata: &'a L1BatchWithMetadata, pubdata_da: PubdataDA) -> Self {
+        Self {
+            l1_batch_with_metadata,
+            pubdata_da,
+        }
+    }
+
+    fn base_tokens(&self) -> Vec<Token> {
```

Comment:
is this the same code as CommitBatchInfoRollup? or does it differ - if it does differ, it should be clearly marked where. If it doesn't - then maybe we should have a single function?